examples/chat/server_test.go

```diff
@@ -0,0 +1,126 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"regexp"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/ServiceWeaver/weaver"
+	"github.com/ServiceWeaver/weaver/weavertest"
+)
+
+// db is a fake store for the chat database.
+type db struct {
+	mu       sync.Mutex
+	lastPost PostID
+	threads  []Thread
+}
+
+var _ SQLStore = &db{}
+
+func (db *db) CreateThread(ctx context.Context, creator string, when time.Time, others []string, text string, image []byte) (ThreadID, error) {
+	db.mu.Lock()
+	defer db.mu.Unlock()
+	db.lastPost++
+	t := Thread{
+		ID:    ThreadID(len(db.threads)) + 1,
+		Posts: []Post{{ID: db.lastPost, Creator: creator, When: when, Text: text}}, // image is ignored
+	}
+	db.threads = append(db.threads, t)
+	return t.ID, nil
+}
+
+func (db *db) CreatePost(ctx context.Context, creator string, when time.Time, thread ThreadID, text string) error {
+	db.mu.Lock()
+	defer db.mu.Unlock()
+	db.lastPost++
+	t := db.threads[int(thread-1)]
+	t.Posts = append(t.Posts, Post{ID: db.lastPost, Creator: creator, When: when, Text: text})
+	db.threads[int(thread-1)] = t
+	return nil
+}
+
+func (db *db) GetFeed(ctx context.Context, user string) ([]Thread, error) {
+	db.mu.Lock()
+	defer db.mu.Unlock()
+	return db.threads, nil
```

Comment:
Do we have to return a deep copy of `db.threads` to avoid races? I guess it doesn't really matter for these tests?

Reply:
Done.